src/Features/CSharp/Portable/CodeRefactorings/AddNamedArguments/CSharpAddNamedArgumentsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,211 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Composition;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CodeRefactorings.AddNamedArguments;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.AddNamedArguments
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(CSharpAddNamedArgumentsCodeRefactoringProvider)), Shared]
+    internal class CSharpAddNamedArgumentsCodeRefactoringProvider : AbstractAddNamedArgumentsCodeRefactoringProvider<ArgumentSyntax>
+    {
+        protected override bool TryGetOrSynthesizeNamedArguments(SeparatedSyntaxList<ArgumentSyntax> arguments, ImmutableArray<IParameterSymbol> parameters, out ArgumentSyntax[] namedArguments, out bool hasLiteral)
+        {
+            hasLiteral = false;
+
+            // Early return when already fixed
+            if (arguments[0].NameColon != null)
+            {
+                namedArguments = default(ArgumentSyntax[]);
+                return false;
+            }
+
+            var result = ArrayBuilder<ArgumentSyntax>.GetInstance(arguments.Count);
+            for (int index = 0; index < arguments.Count; ++index)
+            {
+                var argument = arguments[index];
+                if (argument.NameColon != null)
+                {
+                    result.Add(argument);
+                }
+                else
+                {
+                    var parameter = parameters[index];
+                    var namedArgument = argument.WithNameColon(NameColon(parameter.Name));
+                    if (parameter.IsParams && arguments.Count - index > 1)
+                    {
+                        var arrayBuilder = ArrayBuilder<ExpressionSyntax>.GetInstance(arguments.Count - index);
+                        for (; index < arguments.Count; ++index)
+                        {
+                            arrayBuilder.Add(arguments[index].Expression);
+                        }
+
+                        var expressions = arrayBuilder.ToArrayAndFree();
+                        var initializer = InitializerExpression(SyntaxKind.ArrayInitializerExpression, SeparatedList(expressions));
+                        var arrayType = parameter.Type.GenerateTypeSyntax() as ArrayTypeSyntax;
```

Comment:
I would not do the 'As' here.  First, because the type might not be a parameter type.  Second, because if the as fails, you've now introduced null into the system which will kill other things.  

Also, please add a test where you have something like "Foo(params InvalidType x)" to run through this case.